### PR TITLE
feat: New module `avm/ptn/app/container-job-toolkit`

### PR DESCRIPTION
**Pull request by ReneHezser**
## Description

New pattern module

## Pipeline Reference

| Pipeline |
| -------- |
|  [avm.ptn.app.container-job-toolkit.yml](https://github.com/ReneHezser/bicep-registry-modules/actions/workflows/avm.ptn.app.container-job-toolkit.yml)    |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

**Comment by ReneHezser** [comment]
I've created the two GitHub groups, which need to be added to the parent groups:
https://github.com/orgs/Azure/teams/avm-ptn-app-containerjob-module-owners-bicep
https://github.com/orgs/Azure/teams/avm-ptn-app-containerjob-module-contributors-bicep

**Comment by ReneHezser** [comment]
@ReneHezser 
We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.

Please reach out if any support is needed.

File: .github/workflows/avm.ptn.app.container-job-toolkit.yml
Comment (eriqua) [inline .github/workflows/avm.ptn.app.container-job-toolkit.yml:4]: This looks outdated, the trigger on schedule has been removed

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:1]: This will be the readme title, hence it usually is a short user friendly name, e.g., 
```suggestion
metadata name = 'Container job toolkit'
```

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:58]: This came up already discussing with @AlexanderSehr although we probably did not get to a final statement yet.
I'm curios to hear your thoughts a s well:
Should a Conditional parameter have a default value? We enforce a "Required if" statement for all Conditional parameter descriptions. Is it really "Required if" if we anyway have a default value for it?

File: avm/ptn/app/container-job-toolkit/tests/e2e/max/main.test.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/tests/e2e/max/main.test.bicep:1]: If scenario based, I'd suggest to rename the test folder `max` to something more aligned with the use case, e.g., `existing`.

File: avm/ptn/app/container-job-toolkit/tests/e2e/max/main.test.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/tests/e2e/max/main.test.bicep:18]: If renaming the folder please also rename the serviceshort accordingly

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (AlexanderSehr) [reply avm/ptn/app/container-job-toolkit/main.bicep:58]: I'll hold my thoughts until @ReneHezser gave his thoughts 😄

File: .github/workflows/avm.ptn.app.container-job-toolkit.yml
Comment (ReneHezser) [reply .github/workflows/avm.ptn.app.container-job-toolkit.yml:4]: Thanks. I didn't find any template, so I copied from an existing action. Would you have the template link for me?

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/main.bicep:58]: I understand the concerns. Technically, it is not required if it has a default value.
Since it is relevant only under circumstances, I pointed this out with "Required if", while still offering a default value.
Should we think about e.g. an "Applies if" in addition to a "Required if" to point out the scope of the parameter?

File: avm/ptn/app/container-job-toolkit/tests/e2e/max/main.test.bicep
Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/tests/e2e/max/main.test.bicep:1]: My intention was to show the usage of all parameters and not a single use-case. Therefore, I changed the name. Thanks @eriqua

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (AlexanderSehr) [reply avm/ptn/app/container-job-toolkit/main.bicep:58]: I'd not want to overcomplicate this (by adding yet another case). @eriqua and I come from two different angles when it comes to this matter. @eriqua I'm not sure how we best continus from here. I guess we could discuss it in a call amongst the 3 of us? Or put it up for debate in the larger group? Whatever it is - getting to this 'final statement', you referred to before, would be great to get to and document (in the specs). Let me know you thoughts :)

Comment (eriqua) [reply avm/ptn/app/container-job-toolkit/main.bicep:58]: @AlexanderSehr, @ReneHezser works for me, probably a sync between the 3 of us would be enough. I don't want to block the PR on this, we have the same in many other places, so once discussed and documented we can, if ever, align the implementation as well. I just saw it and thought about mentioning. You can resolve this comment if you agree 🙂

Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/main.bicep:58]: I put this topic on the agenda for today's call.

File: avm/ptn/app/container-job-toolkit/tests/e2e/max/main.test.bicep
Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/tests/e2e/max/main.test.bicep:18]: Not needed (see above comment)

**Comment by ReneHezser** [comment]
@eriqua I updated references and the pipeline is green.

File: .github/CODEOWNERS
Comment (AlexanderSehr) [inline .github/CODEOWNERS:1]: Please note, the `main.json` templates **must** be re-build with the latest Bicep version `0.31.34` as this version (running on the agents) renders a different `main.json` file then previous versions. 

The steps to execute would be
- Upgrade your Bicep CLI version to `0.31.34` (e.g. via `choco upgrade bicep` if choco is used)
- Re-run the `Set-AVMModule` script on your module. E.g. via `Set-AVMModule -ModuleFolderPath '.\avm\res\automation\automation-account\' -SkipFileAndFolderSetup -SkipReadMe -Recurse`
- Push the changes
- Re-run the tests to ensure everything works. 

If the static tests continue to fail for you, please note that there is currently presumably another new bug which affects some of our modules ([ref](https://github.com/Azure/bicep/issues/15517)).

Comment (ReneHezser) [reply .github/CODEOWNERS:1]: Thanks Alex. Done.

**Comment by jtracey93** [comment]
Hey @ReneHezser ,

Firstly, thanks for your work on this PR!

We have made some changes to the AVM CI, detailed below, which means we need you to update your fork to pull in these latest changes and re-run your tests to show they still are passing prior to approving and merging this PR, as we don't and it fails once merged the publishing of your module will fail and will be blocked going forward until the test pass again via additional PRs.

### Changes to CI That Have Been Made That You Need To Take Action On

- https://github.com/Azure/bicep-registry-modules/pull/3745
    - Adds WAF Security PSRule Checks using this custom baseline of rules ([`/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/cb-waf-security.Rule.yaml`](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/cb-waf-security.Rule.yaml))
    - **Action Required By You:**
        - Ingest/pull-in these changes into your PR branch and re-run your tests to ensure they pass
            - If any issues with the rules and you think it should apply, please ping/reach out to the core team

Any questions reach out to the AVM Core Team by tagging us in your PR here or internally via Teams

Thanks

Jack (AVM Core Team)

File: .github/workflows/avm.ptn.app.container-job-toolkit.yml
Comment (eriqua) [inline .github/workflows/avm.ptn.app.container-job-toolkit.yml:60]:
```suggestion
          workflowPath: "${{ env.workflowPath }}"
```

File: avm/ptn/app/container-job-toolkit/tests/e2e/max/main.test.bicep
Comment (eriqua) [reply avm/ptn/app/container-job-toolkit/tests/e2e/max/main.test.bicep:1]: In this case, please align the metadata description for max files `This instance deploys the module with most of its features enabled.`

File: avm/ptn/app/container-job-toolkit/tests/e2e/waf-aligned/main.test.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/tests/e2e/waf-aligned/main.test.bicep:4]: Also in this case I'd suggest to keep the aligned description for waf aligned tests `This instance deploys the module in alignment with the best-practices of the Azure Well-Architected Framework.` If private networking and workload plan are required by WAF it's OK to keep that info after the usual description

File: avm/ptn/app/container-job-toolkit/tests/e2e/waf-aligned/dependencies.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/tests/e2e/waf-aligned/dependencies.bicep:18]: Why using a module instead of a resource block here? Just curious if there is a reason.

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:930]: Please move each import before corresponding resource

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:875]: Please keep UDT singular and not nullable
```suggestion
type securityRuleType = {
```

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:928]: Please keep UDT singular and not nullable
```suggestion
}
```

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:37]: Please keep UDT singular and not nullable
```suggestion
param customNetworkSecurityGroups securityRuleType[]?
```

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:203]: Please leave one empty line before each module declaration for better readability (applies to whole file)

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:108]: Please keep UDT singular
```suggestion
param environmentVariables environmentVariableType[]?
```

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:151]: Is this an array of objects? can this be typed?

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:153]:
```suggestion
@description('Optional. The name of the workload profile to use. Leave empty to use a consumption based profile.')
```

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:318]: Why the condition here only and not added to other conditional outputs?

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:332]:
```suggestion
type environmentVariableType = {
```

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:336]:
```suggestion
  @description('Conditional. The name of the Container App secret from which to pull the environment variable value. Required if `value` is null.')
```

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:339]: What happens if both value and secretRef are provided? If either takes precedence, I'd suggest to specify that in the description

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:332]: Is this type worth exporting?

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/main.bicep:331]: Please add type level `@description()`

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:671]: Please leave one empty line between each block for readability

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:874]: Please add type level `@description()`

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:887]:
```suggestion
    @description('Optional. The destination address prefix. CIDR or destination IP range. Asterisk "*" can also be used to match all source IPs. Default tags such as "VirtualNetwork", "AzureLoadBalancer" and "Internet" can also be used.')
```

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:907]:
```suggestion
    @description('Required. The priority of the rule. The value can be between 100 and 4096. The priority number must be unique for each rule in the collection. The lower the priority number, the higher the priority of the rule.')
```

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:871]: Are we sure we want to allow this?

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:855]: Please add type level `@description()`

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:822]: Please add `@description()` to all outputs

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [inline avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:793]: Is there a reason for not using the module (just double-checking if intended)

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/main.bicep:339]: That is unclear, as https://learn.microsoft.com/en-us/azure/container-apps/containers#configuration doesn't specify the order.

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/main.bicep:151]: A UDT is best suited. Thanks for the hint.

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/main.bicep:318]: Uh, probably copy&paste

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/main.bicep:339]: That is unfortunately unclear, as not specified in the ContainerJob documentation.

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep:871]: I would like to keep it, as there might be cases where Key Vault is not possible or other reasons, that I currently can't think of.

File: avm/ptn/app/container-job-toolkit/tests/e2e/waf-aligned/dependencies.bicep
Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/tests/e2e/waf-aligned/dependencies.bicep:18]: I wanted to use the connectionString output from it

**Comment by ReneHezser** [comment]
Thank you @eriqua 
I've gone through your remarks and ran the pipeline again.

File: avm/ptn/app/container-job-toolkit/main.bicep
Comment (eriqua) [reply avm/ptn/app/container-job-toolkit/main.bicep:329]: Fair enough, just wanted to mention so that you can give it a thought for next PRs. Resolving for this first version

File: avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep
Comment (eriqua) [reply avm/ptn/app/container-job-toolkit/modules/deploy_services.bicep]: Definitely your call. Just wanted to mention. Resolving.

File: avm/ptn/app/container-job-toolkit/tests/e2e/waf-aligned/main.test.bicep
Comment (eriqua) [reply avm/ptn/app/container-job-toolkit/tests/e2e/waf-aligned/main.test.bicep:4]: @ReneHezser thanks for the awesome work and for incorporating all comments. I think this one only is left (probably due to GH UI hiding them). LGTM for the rest. Thanks again!

Comment (ReneHezser) [reply avm/ptn/app/container-job-toolkit/tests/e2e/waf-aligned/main.test.bicep]: @eriqua I thought I had changed the description. Well, now I have.
I also update references and re-ran the pipeline.

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**
